packages/expo-in-app-purchases/ios/EXInAppPurchases/EXInAppPurchasesModule.m

```diff
@@ -0,0 +1,449 @@
+
+UM_EXPORT_MODULE(ExpoInAppPurchases);
+
+- (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
+{
+  _moduleRegistry = moduleRegistry;
+  _eventEmitter = [moduleRegistry getModuleImplementingProtocol:@protocol(UMEventEmitterService)];
+}
+
+- (NSArray<NSString *> *)supportedEvents
+{
+  return @[EXPurchasesUpdatedEventName];
+}
+
+- (void)startObserving {}
+- (void)stopObserving {}
+
+# pragma mark - Exported Methods
+
+UM_EXPORT_METHOD_AS(connectAsync,
+                    connectAsync:(UMPromiseResolveBlock)resolve
+                    reject:(UMPromiseRejectBlock)reject)
+{
+  // Initialize listener and object properties
+  [[SKPaymentQueue defaultQueue] addTransactionObserver:self];
+
+  _promises = [NSMutableDictionary dictionary];
+  _pendingTransactions = [NSMutableDictionary dictionary];
+  _retrievedItems = [NSMutableSet set];
+
+  _queryingItems = NO;
+  BOOL promiseSet = [self setPromise:QUERY_HISTORY_KEY resolve:resolve reject:reject];
+
+  if (promiseSet) {
+    // Request history
+    [[SKPaymentQueue defaultQueue] restoreCompletedTransactions];
+  }
+}
+
+UM_EXPORT_METHOD_AS(getProductsAsync,
+                    getProductsAsync:(NSArray *)productIDs
+                    resolve:(UMPromiseResolveBlock)resolve
+                    reject:(UMPromiseRejectBlock)reject)
+{
+  [self setPromise:QUERY_PURCHASABLE_KEY resolve:resolve reject:reject];
+
+  for (NSString *identifier in productIDs) {
+    [_retrievedItems addObject:identifier];
+  }
+  _queryingItems = YES;
+  [self requestProducts:productIDs];
+}
+
+UM_EXPORT_METHOD_AS(purchaseItemAsync,
+                    purchaseItemAsync:(NSString *)productIdentifier
+                    resolve:(UMPromiseResolveBlock)resolve
+                    reject:(UMPromiseRejectBlock)reject)
+{
+  if (![SKPaymentQueue canMakePayments]) {
+    reject(@"E_MISSING_PERMISSIONS", @"User cannot make payments", nil);
+    return;
+  }
+  if (![_retrievedItems containsObject:productIdentifier]) {
+    reject(@"E_ITEM_NOT_QUERIED", @"Must query item from store before calling purchase", nil);
+    return;
+  }
+
+  // Make the request
+  BOOL promiseSet = [self setPromise:productIdentifier resolve:resolve reject:reject];
+  if (promiseSet) {
+    _queryingItems = NO;
+    [self requestProducts:@[productIdentifier]];
+  }
+}
+
+UM_EXPORT_METHOD_AS(finishTransactionAsync,
+                    finishTransactionAsync:(NSString *)purchaseToken
+                    consume:(BOOL)consume // ignore
+                    resolve:(UMPromiseResolveBlock)resolve
+                    reject:(UMPromiseRejectBlock)reject)
+{
+  SKPaymentTransaction *transaction = _pendingTransactions[purchaseToken];
+  _pendingTransactions[purchaseToken] = nil;
+  if (transaction != nil) {
+    [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
+  }
+  resolve(nil);
+}
+
+UM_EXPORT_METHOD_AS(getPurchaseHistoryAsync,
+                    getPurchaseHistoryAsync:(BOOL)refresh // ignore
```

Comment:
@sjchmiela it's been a minute since I last checked, is `BOOL` supported now or should this be `NSNumber`?

Reply:
Seems to work